dali/util/fits.cc

```diff
@@ -115,11 +131,142 @@ void ParseHeader(HeaderData &parsed_header, fitsfile *src) {
+  }
+
   for (auto it = dims.rbegin(); it != dims.rend(); ++it) {
     parsed_header.shape.shape.push_back(static_cast<int64_t>(*it));
   }
 }
 
+template <unsigned level>
+inline void extract_data(fitsfile* fptr, std::vector<std::vector<uint8_t>>& raw_data,
+                         std::vector<int64_t>& tile_offset, std::vector<int64_t>& tile_size,
+                         int64* ftile, int64* ltile, int64* tilesize, int64* thistilesize,
+                         int64* rowdim, int64* naxis, int64* offset, int* size, int64* sum_nelemll,
+                         int* status, unsigned max_level = level) {
+  for (int i = ftile[level]; i <= ltile[level]; ++i) {
+    // first and last image pixels along each dimension of the compression tile
+    int64 tfpixel = (i - 1) * tilesize[level] + 1;
+    int64 tlpixel = minvalue(tfpixel + tilesize[level] - 1, naxis[level]);
+    if (level == max_level) {
+      thistilesize[level] = tlpixel - tfpixel + 1;
+      offset[level] = (i - 1) * rowdim[level];
+    } else {
+      thistilesize[level] = thistilesize[level + 1] * (tlpixel - tfpixel + 1);
+      offset[level] = (i - 1) * rowdim[level] + offset[level + 1];
+    }
+    extract_data<level - 1>(fptr, raw_data, tile_offset, tile_size, ftile, ltile, tilesize,
+                            thistilesize, rowdim, naxis, offset, size, sum_nelemll, status,
+                            max_level);
+  }
+}
+
+template <>
+inline void extract_data<0>(fitsfile* fptr, std::vector<std::vector<uint8_t>>& raw_data,
+                            std::vector<int64_t>& tile_offset, std::vector<int64_t>& tile_size,
+                            int64* ftile, int64* ltile, int64* tilesize, int64* thistilesize,
+                            int64* rowdim, int64* naxis, int64* offset, int* size,
+                            int64* sum_nelemll, int* status, unsigned max_level) {
+  for (int i = ftile[0]; i <= ltile[0]; ++i) {
+    int64 tfpixel = (i - 1) * tilesize[0] + 1;
+    int64 tlpixel = minvalue(tfpixel + tilesize[0] - 1, naxis[0]);
+    thistilesize[0] = thistilesize[1] * (tlpixel - tfpixel + 1);
+    int64 irow = i + offset[1];
+
+    LONGLONG nelemll = 0, noffset = 0;
+    ffgdesll(fptr, (fptr->Fptr)->cn_compressed, irow, &nelemll, &noffset, status);
+
+    tile_offset[*size] = *sum_nelemll;
+    tile_size[*size] = static_cast<int64>(thistilesize[0]);
+
+    unsigned char charnull = 0;
+    raw_data[*size].resize(nelemll / sizeof(unsigned char));
+    fits_read_col(fptr, TBYTE, (fptr->Fptr)->cn_compressed, irow, 1, static_cast<int64>(nelemll),
+                  &charnull, raw_data[*size].data(), nullptr, status);
+
+    ++(*size);
+    *sum_nelemll += nelemll;
+  }
+}
+
+int extract_undecoded_data(fitsfile* fptr, std::vector<uint8_t>& data,
+                           std::vector<int64_t>& tile_offset, std::vector<int64_t>& tile_size,
+                           int64 rows, int* status) {
```

Comment:
Sure! There might be a bit of a problem with getting grand truths for the tests, since as far as I am concerned, cfitsio doesn't support reading raw uncompressed data without doing compression and scaling first.